lang/ruby/lib/avro/json_io.rb

```diff
@@ -0,0 +1,293 @@
+
+        decoder.read_fixed(writers_schema.size)
+      end
+
+      def read_enum(writers_schema, readers_schema, decoder)
+        validate(readers_schema, decoder.data)
+
+        decoder.data
+      end
+
+      def read_array(writers_schema, readers_schema, decoder)
+        read_items = []
+        decoder.data.each do |item|
+          item_decoder = decoder.class.new
+          item_decoder.data = item
+          read_items << read_data(writers_schema.items, readers_schema.items, item_decoder)
+        end if decoder.data
+
+        validate(readers_schema, read_items)
+
+        read_items
+      end
+
+      def read_map(writers_schema, readers_schema, decoder)
+        read_items = {}
+        decoder.data.each_pair do |key, value|
+          value_decoder = decoder.class.new
+          value_decoder.data = value
+          read_items[key] = read_data(writers_schema.values, readers_schema.values, value_decoder)
+        end
+
+        validate(readers_schema, read_items)
+
+        read_items
+      end
+
+      def read_union(writers_schema, readers_schema, decoder)
+        if decoder.data.nil?
+          schema_type, value = :null, nil
+        else
+          schema_type, value = decoder.data.first
+        end
+
+        data_schema = readers_schema.schemas.find do |reader_schema|
+          reader_schema.type_sym == schema_type.to_sym ||
+                                    (reader_schema.name == schema_type if reader_schema.respond_to?(:name))
+        end
+
+        raise AvroTypeError.new(readers_schema, value) unless data_schema
+
+        value_decoder = decoder.class.new
+        value_decoder.data = value
+        field_value = read_data(data_schema, readers_schema, value_decoder)
+
+        validate(data_schema, field_value)
+
+        field_value
+      end
+    end # JsonDatumReader
+
+    # JsonEncoder encodes a given datum and, when asked, writes it to the writer.
+    class JsonEncoder
+      attr_reader :writer
+      attr_accessor :data
+
+      def initialize(writer)
+        @writer = writer
+      end
+
+      def write
+        writer.write(JSON.dump(@data))
+      end
+
+      def write_null(datum)
+        @data = nil
+      end
+
+      def write_boolean(datum)
+        @data = datum
+      end
+
+      def write_int(n)
+        write_long(n)
+      end
+
+      def write_long(n)
+        @data = Integer(n)
+      end
+
+      def write_float(n)
+        @data = Float(n)
+      end
+
+      def write_double(n)
+        write_float(n)
+      end
+
+      def write_bytes(datum)
+        @data = datum.encode(Encoding::UTF_8, Encoding::ISO_8859_1)
+      end
+
+      def write_fixed(datum)
+        @data = write_bytes(datum)
+      end
+
+      def write_string(datum)
+        @data = datum.encode(Encoding::UTF_8)
+      end
+    end # JsonEncoder
+
+    # Unlike DatumWriter, JsonDatumWriter writes the encoded data in memory. At the end, it writes the data to the writer.
```

Comment:
Nit: we generally try to wrap lines at 80. It's fine to be a little over sometimes, but when the line is too long to display on github I think it should be wrapped.